### ISSUE TITLE
option to backup

### ISSUE THREAD

**Issue by dgc1980** [coollabsio/coolify#233]
I would love an feature to backup current configurations and storage used by databases etc to be generated, or even able to schedule it

**Comment by andrasbacsai** [comment]
Great idea. You would like to download the database backups and the configurations locally? Or you would like to upload directly to for example a S3 compatible place?

**Comment by dgc1980** [comment]
automatic backup of the entire setup would be great, with an option to restore

but also manual backup/restore for relocation/reinstall

**Comment by richmans** [comment]
Some form of automated backup would be great. I would prefer s3 upload, because that removes any need for additional shellscripting. If coolify offers a download, that would mean i'm putting my credentials inside the backupscript, and that's ugly!

**Comment by tiero** [comment]
> automatic backup of the entire setup would be great, with an option to restore

+1

**Comment by vasani-arpit** [comment]
There should be two options one will be automatic based which will be done by coolify automatically to s3 or gDrive based on settings.

another option should be CLI based something like `coolify backup <container name>`  this will create a zip which I can share with others for debugging/development purposes or to keep it on network somewhere else.

**Comment by jankohlbach** [comment]
+1 for the backup option, I really do like coolify, but I kind of need an easy option to make sure I don't lose data for some apps I want to put up there

**Comment by phaberest** [comment]
Sorry, I'm new to coolify and have not yet installed it so I'm just guessing and all I say might be wrong, but:

If it runs the applications via docker it would be possible to share the application folder and run a cronjob to regularly backup the files, could it be a solution?

**Comment by richmans** [comment]
You are not wrong! In most cases that works, although there are a couple of nasty details:

> With some applications such as databases, you cant just copy the files, you have to use a tool to make a dump
> you need to implement some kind of check and notification mechanism to make sure you get notified when the backup fails

None of this is rocketscience and can probably be arranged with a couple of lines of bash. But i think the idea of coolify is that things happen automagically. It would be cool if that included something as crucial as backups.

**Comment by phaberest** [comment]
> * With some applications such as databases, you cant just copy the files, you have to use a tool to make a dump

Yeah, on my current VPSs I'm using a bash script that reads the database user from the `.env` and compresses the folder. It works almost magically until I keep the structure and I have all the websites inside the same folder `/var/www`

I'll give coolify a shot to better understand how it works and will try to sort out a way

**Comment by kpietraszko** [comment]
I think there should be 2 backup types. 
One for coolify config, variables, etc.
And another for databases. This could possible be added as a Service, for Postgres for example you could use something like: https://github.com/schickling/dockerfiles/tree/master/postgres-backup-s3

**Comment by ErSauravAdhikari** [comment]
Hmm, what's the progress with this?

**Comment by jnreis** [comment]
I think there are two good options for this problem.

The first one would be setup up [Restic](https://github.com/restic/restic) during the Coolify installation. The second one would be to add [Docker Volume Backup (by Offen)](https://github.com/offen/docker-volume-backup) to each Application, Service or Database that has volumes.

Then, in the Persistent Volumes tab (or on its own tab) of an Application, Service or Database add options to Backup and Restore the Persistent Volumes. The options may include Scheduled Backups (where multiple schedules can be added; they can be displayed in a table with update and delete options for each one), Manual Backups, and Manual Restore of one or more volumes (a simple checkbox for each volume will suffice). The backups are fetched from the source and also displayed in a table.

Restic on Docker Hub: https://hub.docker.com/u/restic
Offen on Docker Hub: https://hub.docker.com/u/offen

Both options are well stablished and under active development.

I do believe that this feature would be an incredible addition to Coolify's already impressive set of features. And I'm certain that it would hugely benefit Coolify users by giving them (including me) more peace of mind.

**Comment by m1daz** [comment]
As someone who owns a server with 8tb nvme storage and 16tb sata ssd storage... I much rather have my backups on one of the 8 drives than on S3. Please add local backup options

**Comment by itzJustKranker** [comment]
@m1daz couldnt this be achieved with ssh access to server and a simple cron gzip for /data/coolify folder? 

`tar -zcvf "/backups/$(date '+%Y-%m-%d')-coolify-backup.tar.gz" /data/coolify`

@andrasbacsai I don't know enough about how coolify restores, but if I hypothetically had a tar backup of /data/coolify, could I simply restore from it on a completely new server and expect everything to continue working? After running the install.sh of course. It would be nice to offer built in local backup/restore alongside s3 if desired.

**Comment by m1daz** [comment]
> @m1daz couldnt this be achieved with ssh access to server and a simple cron gzip for /data/coolify folder?
> 
> `tar -zcvf "/backups/$(date '+%Y-%m-%d')-coolify-backup.tar.gz" /data/coolify`
> 
> @andrasbacsai I don't know enough about how coolify restores, but if I hypothetically had a tar backup of /data/coolify, could I simply restore from it on a completely new server and expect everything to continue working? After running the install.sh of course. It would be nice to offer built in local backup/restore alongside s3 if desired.

It would be nice to have the software do that itself though, right? I mean, in the same way, you could argue S3 backups shouldn't be implemented in the software because you could zip it and manually upload it etc etc. 

I have 4x 4tb drives and without doing some abstract linux configuration I would like for it to do it by itself every night at 21:00 or something. Like on /mnt/whatever save that zip as you mentioned of specifically the database or instance or whatever.
